{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gl/g6/wlouie/anaconda3/envs/skipguide_data_processing/lib/python3.7/site-packages/ipykernel_launcher.py:28: MatplotlibDeprecationWarning: \n",
      "The examples.directory rcparam was deprecated in Matplotlib 3.0 and will be removed in 3.2. In the future, examples will be found relative to the 'datapath' directory.\n",
      "Using TensorFlow backend.\n",
      "/data/gl/g6/wlouie/anaconda3/envs/skipguide_data_processing/lib/python3.7/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator HuberRegressor from version 0.19.2 when using version 0.20.0. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n",
      "/data/gl/g6/wlouie/anaconda3/envs/skipguide_data_processing/lib/python3.7/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator StandardScaler from version 0.19.2 when using version 0.20.0. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n",
      "/data/gl/g6/wlouie/anaconda3/envs/skipguide_data_processing/lib/python3.7/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator LogisticRegression from version 0.19.2 when using version 0.20.0. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n",
      "/data/gl/g6/wlouie/anaconda3/envs/skipguide_data_processing/lib/python3.7/site-packages/sklearn/base.py:251: UserWarning: Trying to unpickle estimator Pipeline from version 0.19.2 when using version 0.20.0. This might lead to breaking code or invalid results. Use at your own risk.\n",
      "  UserWarning)\n"
     ]
    }
   ],
   "source": [
    "from config import *\n",
    "from utils import *\n",
    "\n",
    "import os\n",
    "import sys\n",
    "import regex\n",
    "import copy\n",
    "import numpy as np\n",
    "import collections\n",
    "import multiprocessing\n",
    "import pickle\n",
    "\n",
    "import numpy as np\n",
    "import scipy\n",
    "\n",
    "# Suppress pandas future warning, which messes tqdm\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import pandas as pd\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "inline_rc = dict(mpl.rcParams)\n",
    "import subprocess\n",
    "\n",
    "import time\n",
    "os.environ['KMP_WARNINGS'] = 'off'\n",
    "\n",
    "from mmsplice.vcf_dataloader import SplicingVCFDataloader\n",
    "from mmsplice import MMSplice, predict_all_table, predict_save\n",
    "from mmsplice.utils import max_varEff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "indel_splice_precas_count_map = load_bc_seq(INDEL_SPLICE_PRECAS_COUNT_MAP)\n",
    "indel_splice_postcas_count_map = load_bc_seq(INDEL_SPLICE_POSTCAS_COUNT_MAP)\n",
    "gt_splice_count_map = load_bc_seq(GT_SPLICE_COUNT_MAP)\n",
    "gt_precas_splice_count_map = load_bc_seq(GT_PRECAS_SPLICE_COUNT_MAP)\n",
    "predicted_gt_indel_dist_map = load_var(PREDICTED_GT_INDEL_DIST_MAP)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MMSplice Predict\n",
    "MMSplice requires as input a reference genome FASTA file, a genome annotation file in the standard GTF format, and a variant calling format file (VCF). To represent a query repair genotype as these formats, we considered our lib-SA sequence with the corresponding WT target as the reference genome, and the repair genotype as the variant compared to the WT. That is, the FASTA file contained our lib-SA sequence with the corresponding WT target sequence, the GTF file contained the exonic annotations, and the VCF described the indel of the WT target that resulted in our repair genotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXON_A = 'CAAGATCCGCCACAACATCGAG'\n",
    "INTRON_TARGETSTART = 'GTAAGTTATCACCTTCGTGGCTACAGAGTTTCCTTATTTGTCTCTGTTGCCGGCTTATATGGACAAGCATATCACAGCCATTTATCGGAGCGCCTCCGTACACGCTATTATCGGACGCCTCGCGAGATCAATACGATTACCAGCTGCCCTCGTCGAC'\n",
    "TARGETEND_EXON_B = 'TGATTACACATATAGACACGCGAGCAGCCATCTTTTATAGAATGGGTAGAACCCGTCCTAAGGACTCAGATTGAGCATCGTTTGCTTCTCGAGTACTACCTGGTACAGATGTCTCTTCAAACAG'\n",
    "EXON_C_BCSTART = 'GACGGCAGCGTGCAGCTCGCC'\n",
    "BCEND_EXON_C = 'GACCACTACCAGCAGAACACCCC'\n",
    "\n",
    "TARGET_PREFIX = EXON_A + INTRON_TARGETSTART\n",
    "TARGET_SUFFIX_BC_PREFIX = TARGETEND_EXON_B + EXON_C_BCSTART\n",
    "BC_SUFFIX = BCEND_EXON_C\n",
    "\n",
    "\n",
    "def construct_genome(indel_splice_count_map, outputfn=''):\n",
    "    with open(outputfn, 'w') as f:\n",
    "        for i, (indel, splice_count_map) in enumerate(indel_iterator(indel_splice_count_map)):\n",
    "            target = exp_tid_target_map[exp_gid_tid_map[splice_count_map['gid']]]\n",
    "            genome = TARGET_PREFIX + target + TARGET_SUFFIX_BC_PREFIX\n",
    "            f.write('>' + str(i) + '\\n')\n",
    "            f.write(genome + '\\n')\n",
    "        \n",
    "            \n",
    "def construct_gtf(indel_splice_count_map, outputfn=''):\n",
    "    with open(outputfn, 'w') as f:\n",
    "        for i, (indel, splice_count_map) in enumerate(indel_iterator(indel_splice_count_map)):\n",
    "            target = exp_tid_target_map[exp_gid_tid_map[splice_count_map['gid']]]\n",
    "            genome_len = len(TARGET_PREFIX) + len(target) + len(TARGET_SUFFIX_BC_PREFIX)\n",
    "            spliceidx = 37\n",
    "\n",
    "            # Gene\n",
    "            f.write('\\t'.join([\n",
    "                str(i), 'artificial', 'gene', '1', str(genome_len), '.', '+', '.',\n",
    "                'gene_id \"g{0}\"; transcript_id \"\"; gene_name \"g{0}\";\\n'.format(i)\n",
    "            ]))\n",
    "            # Transcript A_B_C\n",
    "            f.write('\\t'.join([\n",
    "                str(i), 'artificial', 'transcript', '1', str(genome_len), '.', '+', '.',\n",
    "                'gene_id \"g{0}\"; transcript_id \"t{0}.{1}\"; gene_name \"g{0}\";\\n'.format(i, 'ABC')\n",
    "            ]))\n",
    "            # Exon A\n",
    "            f.write('\\t'.join([\n",
    "                str(i), 'artificial', 'exon', '1', str(len(EXON_A)), '.', '+', '.',\n",
    "                'gene_id \"g{0}\"; transcript_id \"t{0}.{1}\"; gene_name \"g{0}\"; exon_id \"e{0}A\";\\n'.format(i, 'ABC')\n",
    "            ]))\n",
    "            # Exon BC\n",
    "            f.write('\\t'.join([\n",
    "                str(i), 'artificial', 'exon', str(len(TARGET_PREFIX) + spliceidx + 1), str(genome_len), '.', '+', '.',\n",
    "                'gene_id \"g{0}\"; transcript_id \"t{0}.{1}\"; gene_name \"g{0}\"; exon_id \"e{0}BC\";\\n'.format(i, 'ABC')\n",
    "            ]))\n",
    "\n",
    "            # Transcript A_C\n",
    "            f.write('\\t'.join([\n",
    "                str(i), 'artificial', 'transcript', '1', str(genome_len), '.', '+', '.',\n",
    "                'gene_id \"g{0}\"; transcript_id \"t{0}.{1}\"; gene_name \"g{0}\";\\n'.format(i, 'AC')\n",
    "            ]))\n",
    "            # Exon A\n",
    "            f.write('\\t'.join([\n",
    "                str(i), 'artificial', 'exon', '1', str(len(EXON_A)), '.', '+', '.',\n",
    "                'gene_id \"g{0}\"; transcript_id \"t{0}.{1}\"; gene_name \"g{0}\"; exon_id \"e{0}A\";\\n'.format(i, 'AC')\n",
    "            ]))\n",
    "            # Exon C\n",
    "            f.write('\\t'.join([\n",
    "                str(i), 'artificial', 'exon', str(len(TARGET_PREFIX) + len(target) + len(TARGETEND_EXON_B) + 1), str(genome_len), '.', '+', '.',\n",
    "                'gene_id \"g{0}\"; transcript_id \"t{0}.{1}\"; gene_name \"g{0}\"; exon_id \"e{0}C\";\\n'.format(i, 'AC')\n",
    "            ]))\n",
    "\n",
    "            \n",
    "def construct_vcf(indel_splice_count_map, outputfn=''):        \n",
    "    indel_splice_count_map_list = list(indel_iterator(indel_splice_count_map))\n",
    "    with open(outputfn, 'w') as f:\n",
    "        f.write('##fileformat=VCFv4.0\\n')\n",
    "        for i in range(len(indel_splice_count_map_list)):\n",
    "            indel, splice_count_map = indel_splice_count_map_list[i]\n",
    "            target = exp_tid_target_map[exp_gid_tid_map[splice_count_map['gid']]]\n",
    "            genome_len = len(TARGET_PREFIX) + len(target) + len(TARGET_SUFFIX_BC_PREFIX)\n",
    "            spliceidx = 37 # Designed splice idx\n",
    "            f.write('##contig=<ID={0},length={1}>\\n'.format(i, genome_len))\n",
    "\n",
    "        f.write('\\t'.join(['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO\\n']))\n",
    "        for i, (indel, splice_count_map) in enumerate(indel_splice_count_map_list):\n",
    "            # Exon B acceptor\n",
    "            if indel[1] == 'N':\n",
    "                pos = len(TARGET_PREFIX) + spliceidx + 1\n",
    "                ref = target[spliceidx]\n",
    "                alt = target[spliceidx]\n",
    "            elif indel[1] in DELETION_SIGNATURES:\n",
    "                _, _, deletion_size, genotype_pos, cutsite = indel\n",
    "                pos = len(TARGET_PREFIX) + cutsite + genotype_pos - deletion_size\n",
    "                ref = target[(cutsite + genotype_pos - deletion_size - 1):(cutsite + genotype_pos + 1)]\n",
    "                alt = target[cutsite + genotype_pos - deletion_size - 1]\n",
    "            elif indel[1] in INSERTION_SIGNATURES and indel[2] == 1:\n",
    "                _, _, _, inserted_base, cutsite = indel\n",
    "                pos = len(TARGET_PREFIX) + cutsite\n",
    "                ref = target[cutsite - 1]\n",
    "                alt = target[cutsite - 1] + inserted_base \n",
    "            else:\n",
    "                continue\n",
    "                \n",
    "            f.write('\\t'.join([\n",
    "                str(i), str(pos), '.', ref, alt, '.', '.', '.', '\\n'\n",
    "            ]))\n",
    "            \n",
    "            \n",
    "def predict_MMSpliceScores(indel_splice_count_map, outfn='MESC_reporter', progress=False):\n",
    "    fasta = os.path.join(TEMP_DIR, outfn + '.fa')\n",
    "    gtf = os.path.join(TEMP_DIR, outfn + '.gtf')\n",
    "    vcf = os.path.join(TEMP_DIR, outfn + '.vcf')\n",
    "    \n",
    "    construct_genome(indel_splice_count_map, fasta)\n",
    "    construct_gtf(indel_splice_count_map, gtf)\n",
    "    construct_vcf(indel_splice_count_map, vcf)\n",
    "    \n",
    "    start = time.perf_counter()\n",
    "    dl = SplicingVCFDataloader(gtf, fasta, vcf)\n",
    "    if progress:\n",
    "        print(time.perf_counter() - start)\n",
    "    \n",
    "    start = time.perf_counter()\n",
    "    model = MMSplice()\n",
    "    if progress:\n",
    "        print(time.perf_counter() - start)\n",
    "    \n",
    "    start = time.perf_counter()\n",
    "    predictions = predict_all_table(model, dl, progress=False, pathogenicity=False, splicing_efficiency=False)\n",
    "    if progress:\n",
    "        print(time.perf_counter() - start)\n",
    "    return predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(MMSPLICE_PRECAS_DF_PATH):\n",
    "    mmsplice_precas_df = predict_MMSpliceScores(indel_splice_precas_count_map, 'MESC_reporter_precas', True)\n",
    "    mmsplice_precas_df.to_csv(MMSPLICE_PRECAS_DF_PATH, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(MMSPLICE_POSTCAS_DF_PATH):\n",
    "    mmsplice_postcas_df = predict_MMSpliceScores(indel_splice_postcas_count_map, 'MESC_reporter_postcas', True)\n",
    "    mmsplice_postcas_df.to_csv(MMSPLICE_POSTCAS_DF_PATH, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_mmsplice_predict_gt_indel_dist(pairs):\n",
    "    for i, p in enumerate(tqdm(pairs)):\n",
    "        gid = exp_grna_gid_map[p[0]][0]\n",
    "        cutsite = get_cutsite(*p)\n",
    "        indel_splice_count_map = collections.defaultdict(lambda: collections.defaultdict(int))\n",
    "        distribution = predicted_gt_indel_dist_map[p]\n",
    "        for deletion_size in range(1, MAX_INDEL_LEN + 1):\n",
    "            for genotype_pos in distribution[-deletion_size]:\n",
    "                if distribution[-deletion_size][genotype_pos] > 0:\n",
    "                    indel = ('', 'DS', deletion_size, genotype_pos, cutsite)\n",
    "                    indel_splice_count_map[indel]['gid'] = gid\n",
    "                    indel_splice_count_map[indel]['indelphifreq'] = distribution[-deletion_size][genotype_pos]\n",
    "        for base in 'AGTC':\n",
    "            if distribution[1][base] > 0:\n",
    "                indel = ('', 'IS', 1, base, cutsite)\n",
    "                indel_splice_count_map[indel]['gid'] = gid\n",
    "                indel_splice_count_map[indel]['indelphifreq'] = distribution[1][base]\n",
    "        \n",
    "        mmsplice_df = predict_MMSpliceScores(indel_splice_count_map, 'MESC_agg_postcas')\n",
    "        mmsplice_df.to_csv(os.path.join(MMSPLICE_GT_DF_DIR, str(gid) + '.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "if not os.path.exists(os.path.join(MMSPLICE_GT_DF_DIR, '626.csv')):\n",
    "    lib_pairs = [gid_to_gt(gid) for gid in exp_gid_tid_map]\n",
    "    save_mmsplice_predict_gt_indel_dist(lib_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:skipguide_data_processing] *",
   "language": "python",
   "name": "conda-env-skipguide_data_processing-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
