{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from config import *\n",
    "from utils import *\n",
    "\n",
    "import os\n",
    "import csv\n",
    "import regex\n",
    "import numpy as np\n",
    "\n",
    "# Suppress pandas future warning, which messes tqdm\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n",
    "import pandas as pd\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "from Bio import SeqIO\n",
    "\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# lib-SA Library\n",
    "Load our designed lib-SA library. There are 1,927 intron-exon sequences, and 4,000 gRNAs that target them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Azimuth score</th>\n",
       "      <th>Cas type</th>\n",
       "      <th>Cutsite within sequence</th>\n",
       "      <th>Designed 61-bp target site (37i-24e, AG)</th>\n",
       "      <th>Designed 61-bp target site (37i-24e, CT reverse-complement)</th>\n",
       "      <th>Designed gRNA (NGG orientation, 19 and 20)</th>\n",
       "      <th>Has AG</th>\n",
       "      <th>MaxEntScan AG disruption frequency in MH dels and insertions genomic_context</th>\n",
       "      <th>MaxEntScan AG disruption frequency in MH dels and insertions library_context</th>\n",
       "      <th>Ontarget score</th>\n",
       "      <th>...</th>\n",
       "      <th>inDelphi Precision library_context</th>\n",
       "      <th>inDelphi expected indel length genomic_context</th>\n",
       "      <th>inDelphi expected indel length library_context</th>\n",
       "      <th>inDelphi highest outcome frequency genomic_context</th>\n",
       "      <th>inDelphi highest outcome frequency library_context</th>\n",
       "      <th>inDelphi logphi genomic_context</th>\n",
       "      <th>inDelphi logphi library_context</th>\n",
       "      <th>strand</th>\n",
       "      <th>total_vars</th>\n",
       "      <th>Designed gRNA (NGG orientation)</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Identifier number</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.685572</td>\n",
       "      <td>SpCas9</td>\n",
       "      <td>68</td>\n",
       "      <td>GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...</td>\n",
       "      <td>GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...</td>\n",
       "      <td>ATCCAGGTCAGGACTGCAGG</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.796188</td>\n",
       "      <td>0.796141</td>\n",
       "      <td>0.858865</td>\n",
       "      <td>...</td>\n",
       "      <td>0.426247</td>\n",
       "      <td>7.552535</td>\n",
       "      <td>7.546523</td>\n",
       "      <td>16.796109</td>\n",
       "      <td>16.798237</td>\n",
       "      <td>-2.441151</td>\n",
       "      <td>-2.441337</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>GATCCAGGTCAGGACTGCAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.666555</td>\n",
       "      <td>SpCas9</td>\n",
       "      <td>69</td>\n",
       "      <td>GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...</td>\n",
       "      <td>GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...</td>\n",
       "      <td>ATCCAGGTCAGGACTGCAG</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.760537</td>\n",
       "      <td>0.760490</td>\n",
       "      <td>0.844692</td>\n",
       "      <td>...</td>\n",
       "      <td>0.527270</td>\n",
       "      <td>8.262278</td>\n",
       "      <td>8.256890</td>\n",
       "      <td>37.627192</td>\n",
       "      <td>37.633988</td>\n",
       "      <td>-2.109904</td>\n",
       "      <td>-2.110079</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>TGATCCAGGTCAGGACTGCAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.621866</td>\n",
       "      <td>SpCas9</td>\n",
       "      <td>70</td>\n",
       "      <td>GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...</td>\n",
       "      <td>GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...</td>\n",
       "      <td>TGATCCAGGTCAGGACTGCA</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.793351</td>\n",
       "      <td>0.793299</td>\n",
       "      <td>0.854025</td>\n",
       "      <td>...</td>\n",
       "      <td>0.572962</td>\n",
       "      <td>6.545344</td>\n",
       "      <td>6.540914</td>\n",
       "      <td>28.429830</td>\n",
       "      <td>28.436377</td>\n",
       "      <td>-2.139149</td>\n",
       "      <td>-2.139360</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>ATGATCCAGGTCAGGACTGCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0.434994</td>\n",
       "      <td>SpCas9</td>\n",
       "      <td>71</td>\n",
       "      <td>GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...</td>\n",
       "      <td>GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...</td>\n",
       "      <td>ATGATCCAGGTCAGGACTGC</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.699063</td>\n",
       "      <td>0.698923</td>\n",
       "      <td>0.781274</td>\n",
       "      <td>...</td>\n",
       "      <td>0.523115</td>\n",
       "      <td>8.410657</td>\n",
       "      <td>8.406734</td>\n",
       "      <td>37.406834</td>\n",
       "      <td>37.408835</td>\n",
       "      <td>-2.126282</td>\n",
       "      <td>-2.126381</td>\n",
       "      <td>+</td>\n",
       "      <td>12</td>\n",
       "      <td>GATGATCCAGGTCAGGACTGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>0.535781</td>\n",
       "      <td>SpCas9</td>\n",
       "      <td>68</td>\n",
       "      <td>TGCCAGGCATGCCCACTGACTGGTGCCCTTGGCCCAGGGTGGAGAC...</td>\n",
       "      <td>AGTATTGGAATGAAGGTCTCCACCCTGGGCCAAGGGCACCAGTCAG...</td>\n",
       "      <td>TGACTGGTGCCCTTGGCCCA</td>\n",
       "      <td>yes</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.860908</td>\n",
       "      <td>...</td>\n",
       "      <td>0.443413</td>\n",
       "      <td>7.775675</td>\n",
       "      <td>7.788979</td>\n",
       "      <td>10.368532</td>\n",
       "      <td>10.361942</td>\n",
       "      <td>-2.355018</td>\n",
       "      <td>-2.354383</td>\n",
       "      <td>+</td>\n",
       "      <td>14</td>\n",
       "      <td>CTGACTGGTGCCCTTGGCCCA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Azimuth score Cas type  Cutsite within sequence  \\\n",
       "Identifier number                                                    \n",
       "2                       0.685572   SpCas9                       68   \n",
       "3                       0.666555   SpCas9                       69   \n",
       "4                       0.621866   SpCas9                       70   \n",
       "5                       0.434994   SpCas9                       71   \n",
       "6                       0.535781   SpCas9                       68   \n",
       "\n",
       "                            Designed 61-bp target site (37i-24e, AG)  \\\n",
       "Identifier number                                                      \n",
       "2                  GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...   \n",
       "3                  GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...   \n",
       "4                  GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...   \n",
       "5                  GGAAATGCCCTGCCTAAGCTTTGGGTGCCCCCTGCAGTCCTGACCT...   \n",
       "6                  TGCCAGGCATGCCCACTGACTGGTGCCCTTGGCCCAGGGTGGAGAC...   \n",
       "\n",
       "                  Designed 61-bp target site (37i-24e, CT reverse-complement)  \\\n",
       "Identifier number                                                               \n",
       "2                  GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...            \n",
       "3                  GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...            \n",
       "4                  GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...            \n",
       "5                  GCAGGATGATGATCCAGGTCAGGACTGCAGGGGGCACCCAAAGCTT...            \n",
       "6                  AGTATTGGAATGAAGGTCTCCACCCTGGGCCAAGGGCACCAGTCAG...            \n",
       "\n",
       "                  Designed gRNA (NGG orientation, 19 and 20) Has AG  \\\n",
       "Identifier number                                                     \n",
       "2                                       ATCCAGGTCAGGACTGCAGG    yes   \n",
       "3                                        ATCCAGGTCAGGACTGCAG    yes   \n",
       "4                                       TGATCCAGGTCAGGACTGCA    yes   \n",
       "5                                       ATGATCCAGGTCAGGACTGC    yes   \n",
       "6                                       TGACTGGTGCCCTTGGCCCA    yes   \n",
       "\n",
       "                   MaxEntScan AG disruption frequency in MH dels and insertions genomic_context  \\\n",
       "Identifier number                                                                                 \n",
       "2                                                           0.796188                              \n",
       "3                                                           0.760537                              \n",
       "4                                                           0.793351                              \n",
       "5                                                           0.699063                              \n",
       "6                                                           1.000000                              \n",
       "\n",
       "                   MaxEntScan AG disruption frequency in MH dels and insertions library_context  \\\n",
       "Identifier number                                                                                 \n",
       "2                                                           0.796141                              \n",
       "3                                                           0.760490                              \n",
       "4                                                           0.793299                              \n",
       "5                                                           0.698923                              \n",
       "6                                                           1.000000                              \n",
       "\n",
       "                   Ontarget score               ...                 \\\n",
       "Identifier number                               ...                  \n",
       "2                        0.858865               ...                  \n",
       "3                        0.844692               ...                  \n",
       "4                        0.854025               ...                  \n",
       "5                        0.781274               ...                  \n",
       "6                        0.860908               ...                  \n",
       "\n",
       "                  inDelphi Precision library_context  \\\n",
       "Identifier number                                      \n",
       "2                                           0.426247   \n",
       "3                                           0.527270   \n",
       "4                                           0.572962   \n",
       "5                                           0.523115   \n",
       "6                                           0.443413   \n",
       "\n",
       "                  inDelphi expected indel length genomic_context  \\\n",
       "Identifier number                                                  \n",
       "2                                                       7.552535   \n",
       "3                                                       8.262278   \n",
       "4                                                       6.545344   \n",
       "5                                                       8.410657   \n",
       "6                                                       7.775675   \n",
       "\n",
       "                   inDelphi expected indel length library_context  \\\n",
       "Identifier number                                                   \n",
       "2                                                        7.546523   \n",
       "3                                                        8.256890   \n",
       "4                                                        6.540914   \n",
       "5                                                        8.406734   \n",
       "6                                                        7.788979   \n",
       "\n",
       "                  inDelphi highest outcome frequency genomic_context  \\\n",
       "Identifier number                                                      \n",
       "2                                                          16.796109   \n",
       "3                                                          37.627192   \n",
       "4                                                          28.429830   \n",
       "5                                                          37.406834   \n",
       "6                                                          10.368532   \n",
       "\n",
       "                   inDelphi highest outcome frequency library_context  \\\n",
       "Identifier number                                                       \n",
       "2                                                          16.798237    \n",
       "3                                                          37.633988    \n",
       "4                                                          28.436377    \n",
       "5                                                          37.408835    \n",
       "6                                                          10.361942    \n",
       "\n",
       "                  inDelphi logphi genomic_context  \\\n",
       "Identifier number                                   \n",
       "2                                       -2.441151   \n",
       "3                                       -2.109904   \n",
       "4                                       -2.139149   \n",
       "5                                       -2.126282   \n",
       "6                                       -2.355018   \n",
       "\n",
       "                   inDelphi logphi library_context strand total_vars  \\\n",
       "Identifier number                                                      \n",
       "2                                        -2.441337      +         12   \n",
       "3                                        -2.110079      +         12   \n",
       "4                                        -2.139360      +         12   \n",
       "5                                        -2.126381      +         12   \n",
       "6                                        -2.354383      +         14   \n",
       "\n",
       "                   Designed gRNA (NGG orientation)  \n",
       "Identifier number                                   \n",
       "2                            GATCCAGGTCAGGACTGCAGG  \n",
       "3                            TGATCCAGGTCAGGACTGCAG  \n",
       "4                            ATGATCCAGGTCAGGACTGCA  \n",
       "5                            GATGATCCAGGTCAGGACTGC  \n",
       "6                            CTGACTGGTGCCCTTGGCCCA  \n",
       "\n",
       "[5 rows x 34 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_design.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Identifier numbers: 4000\n",
      "Designed 61-bp target site (37i-24e, AG)      1927\n",
      "Designed gRNA (NGG orientation, 19 and 20)    4000\n",
      "dtype: int64\n",
      "20    3016\n",
      "19     984\n",
      "Name: Designed gRNA (NGG orientation, 19 and 20), dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(\"Identifier numbers:\", len(exp_design.index))\n",
    "print(exp_design[['Designed 61-bp target site (37i-24e, AG)', 'Designed gRNA (NGG orientation, 19 and 20)']].nunique())\n",
    "print(exp_design['Designed gRNA (NGG orientation, 19 and 20)'].apply(lambda s: len(s)).value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are 4,000 gRNA that target 1,927 splice acceptors, so some splice acceptors have multiple gRNAs that target them. Here's a distribution of the number of gRNAs that target a splice acceptor:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZcAAAEWCAYAAACqitpwAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3debwcVZn/8c+XhCWsAXLBkIWwRBQ3xABxcEHiAogmgzCCCBGjcUEF0RF0GIFBBxwFlFFQJA4BDYgoJiIqGBblJwQTdkQkYEhiQhK2EFYJPL8/zrmk0unu2/fe6tu3L9/369Wv7jpVdeqp6u56+pyqrlJEYGZmVqb1Wh2AmZkNPE4uZmZWOicXMzMrnZOLmZmVzsnFzMxK5+RiZmal69fJRdL3Jf1nSXWNlvSkpEF5+DpJHyuj7lzfbyRNLqu+biz3a5IelvRQXy+7Io59JC1u4fL/VdKi/B6/sVVxNIukkLRzq+MY6Irbucz9z8tRy5KLpAWSnpG0StLjkv4k6ZOSXoopIj4ZEac2WNc7600TEQsjYtOIeKGE2E+W9OOK+vePiOm9rbubcYwCvgDsGhGv6Mtl90PfAj6T3+NbuzOjpI9IeiEnpick3S7pwML4MXmn8+uK+X4s6eSKsh0kvSjpnJ6uSBN++NT9frT6h0GOod8lz0b3P+2smdu91S2X90XEZsD2wOnA8cC0shciaXDZdfYT2wOPRMTyVgdSph6+X9sDd/disTdGxKbAUOAc4BJJQyumGS9p7y7qORJ4DDhU0oa9iKdtDODvl9XR5fseES15AAuAd1aU7Qm8CLw2D18AfC2/HgZcATwOPAr8kZQcL8rzPAM8CXwJGAMEMAVYCPyhUDY413cdcBpwM7ASmAlslcftAyyuFi+wH/BP4Pm8vNsL9X0sv14POBF4EFgOXAhskcd1xjE5x/Yw8B91ttMWef4Vub4Tc/3vzOv8Yo7jgirz7gMsJrVulgNLgaMK41+KOQ9/BLihMBzAp4H7gFXAqcBOwI3AE8ClwAYVy/pKXqcFwOGFujYktS4WAsuA7wNDKuY9HngIuKjKulTdprneJ3OsTwH319iO7wbuze/1OcD1hfercr03zvXtUfGeHQ9cW5jux8DJFcu5H/hUXseDC+UCzsqxrwTuIH/OK+b/OvAC8Gxer+8W3otP5vfiMeB7gPK4nYBrgEfytv8JMDSPW+f7UbG8TVj7c/QksB3pu3gj6fu2FPhu53tdiOfoHM/fu9rGefxHgXty/L8Dts/lfyi8f08CH6yyXWquY5Vpa25r0j7l+8DVpM/09Z1xFNZr58r9Tx6eCNxG+uzfD+xX+I5Oy9vpH8DXgEE1Yutqu74mx/Zo/gx9JZcPIn237s9xzwNG5XGvKsxzL/Bvhfpqrm+t7Q58HJif65sFbFfvfa+57+pNgujNgyrJJZcvBD5V+eaSEsH3gfXz462s+XKtVRdrdgYXkr48Q6ieXP4BvDZP83Pgx8WdXa14gZM7py2Mv441O6uP5jdnR2BT4BfkHWYhjh/muN4APAe8usZ2upCU+DbL8/4NmFIrzop59wFWA/+Vt9kBwNPAlpUx5+GPsG5ymQVsTvrQPwfMzuu1BfAXYHLFss4k7fDfTvrQ7pLHfzvXtVVel18Bp1XM+40875Aq61Jzm1buFKrMO4y0QzgIGAwcQ/pxsE5yIX2Jjyb9gNim4j3blPSZ6fwcrJVcSJ/J54Atgf8FZhXGvYe0QxhK2vm9GhheI9613pfC+l2R5x9N+rHRuXPbGXhX3nYdpJ3Gt7v6rlV8Tio/728CxuftNYaUFI6tiOfq/H4OaWAbT8rv36vz+BOBPzXy/jWyjhXT1tzWpH3KKuBtua7vsO5nfp3kQkoKK3MM6wEjgFflcb8EfkDaj2xD+sH6iRqx1dyupO/FUtKPwY3y8F553L8DdwK75HV6A7B1XuYi4Khc5+6k5Pua7q5vHt43z797nv5/gT/Uet/r7uPrjWzmg9rJ5SbyL/mKN/e/SDvZdT6AlXWxZmewY5WyYnI5vTB+V9IOZRC9Ty6zgU8Xxu1C+qJ1fqACGFkYfzNwaJX1GkTaWe1aKPsEcF2tnUKVncYzneucy5YD4ytjzsMfqfLB27swPA84vjB8BvkLzpoEsUlh/KXAf5K+DE8BOxXGvZk1v3j3ydt+ozrrUnObVvuSVMx7JKnbq3NYpC9kMbmsJv2afD5vs+Kvv5c+O6SW3E25vDK5nA/8srB+z7MmQe1L+mEwHlivi+/GWu9LYf3eUrFtT6gx/yTg1q6+axWfk5qfozzNscDlFfHs241t/Bvyj6I8vB7ph872Xb1/jaxjxbia25q0T7mkMLwpqaU4qjIO1t7//AA4q8qytiV9R4cUyg6j0MJtdLvm+Wqt073AxCrlHwT+WFH2A+Ck7q5vHp4G/E/F9M8DY6q97/UerT7mUs0IUnOs0jdJv3yukvSApBMaqGtRN8Y/SPp1P6yhKOvbLtdXrHsw6YPYqXh219OkN7HSMGCDKnWN6EYsj0TE6gaWVcuywutnqgwX63osIp4qDD9I2hYdpK6mefnkjceB3+byTisi4tk6cTSyTevN+9J7HelbUnkA+6aIGEpqdcwitUKq+SGwraT3FQslDQEOIXXXEBE3klrhH8rD15C6QL4HLJN0nqTNG4i9qOpnRtI2ki6R9A9JT5CSXq8+x5JeKekKSQ/lOv+7Sp3F709X23h74DuF9/9RUgJq6LPcnXVsYFsX43wyx7JdFyGMInVJVdqetN9YWli3H5BaMNXWo952rbWMrpa/V+ey8/IPB4on+HRnfdf6nuXpH2Ht96mr/SrQ+gP6a5G0B2klbqgcFxGrIuILEbEj8D7gOEkTOkfXqLJWeadRhdejSRn6YdKv7I0LcQ1i7R1hV/UuIb3pxbpXs/aOuREP55gq6/pHN+upZa31ZO0PZE9sKWmTwvBo0rZ4mJSIXhMRQ/Nji0gH0Ds1c5suBUZ2DkhScbgof5k+DRxR7ZTmiHgeOIV0/EmFUf9K6j48J+84HiJ9lo8szHt2RLyJ1MX4SlJXR9UwGlinotPyPK+PiM2BD1fE1lV91cafC/wVGJvr/EpFnZXzdbWNF5G6ioYWHkMi4k9dxNapq3VcO7D62/ql772kTUldPEu6WP4i0nGfauXPAcMK67V5RLymRj31tmutZXS1/OsrtuumEfGpwjTdWd+1vmf5+7w1a+9zGvp89ovkImnzfOrnJaTupjurTHOgpJ3zh/YJUtOu87TiZaS++O76sKRdJW1M6na7LNKpyn8DNpL0Xknrk/qHi2f+LAPGFE+brnAx8Pl8WuqmpF8nP61oQXQpx3Ip8HVJm0naHjiO9KutDLcBB0naOJ+OOKWEOk+RtIGktwIHAj+LiBdJv/jPkrQNgKQRkt7TjXp7s01/DbxO0qR8hsvR1EmkEfEIqYvrqzUmuYj0edivUDYZ+BHwOmC3/Ngb2E3S6yTtIWmv/Hl6inTAvtZp8d39PG9GOiD7uKQRrJu0uqpvGbC1pC0q6nwCeFLSq0gnKdTT1Tb+PvBlSa8BkLSFpEO6EWNX6/iSBrb1AZLeImkD0o+EORHR1a/xacBRkiZIWi9/fl8VEUuBq4Az8n5sPUk7SXp7nfWotV2vAF4h6VhJG+bv/F553PnAqZLGKnm9pK3zPK+UdISk9fNjD0mvbnB9K7f7jLyeu+WzHf87T7+gi+2zjlYnl19JWkXKvv9BOhh8VI1pxwK/J33AbgTOiYjr8rjTgBNzs/CL3Vj+RaQ+yYdIB9A+BxARK0m/Xs8nZeynWLuJ/7P8/IikW6rU+6Nc9x+Av5M+3J/tRlxFn83Lf4DUopuR6y/DWaRjHcuA6eQunV54iHQm0JJc1ycj4q953PGkbs2bcnfA70nHTRrV420aEQ+Tuqz+h9TE3xWYS/rFWcu3SV/K11ep7wXgJNIvQPLObgLp+NNDhcc8UvffZFKr5oek7fNgjuNbNZb9HeBgSY9JOruBVTyFdAB2JWkn/4uK8XW/H/k9uhh4IE+zHfBFUpfeqhz3T+sF0NU2jojLSSdsXJLf/7uA/QtVnAxMz8v/tx6sY1FX23oG6f17lHSA/fB665bjv5m0bzorx3A9a37hH0nqvv5LXuZlwPAaVdXcrhGxinTCwPtI36X7gHfk0WeSfmheRUpO00jHeVaRztI7lPS9e4g1J8Y0sr4nU9juETGbdJz056TW6E657m7rPNvK7GUjtzgXk06VvrbV8QxE/XUbS7qAdPLCia2OpS+0cn1b3XIx6xOS3iNpaG7qd/Zz39TisAYUb2MrcnKxl4s3k862eZjU7TApIp5pbUgDjrexvcTdYmZmVjq3XMzMrHRtfcG5YcOGxZgxY1odhplZW5k3b97DEdHR9ZQ919bJZcyYMcydO7fVYZiZtRVJD3Y9Ve+4W8zMzErn5GJmZqVzcjEzs9I5uZiZWemcXMzMrHROLmZmVjonFzMzK52Ti5mZlc7JxczMStfW/9DvjRlzFvZq/g/tNbqkSMzMBh63XMzMrHROLmZmVjonFzMzK52Ti5mZla6pyUXS5yXdLekuSRdL2kjSDpLmSLpP0k8lbZCn3TAPz8/jxzQzNjMza56mJRdJI4DPAeMi4rXAIOBQ4BvAWRExFngMmJJnmQI8FhE7A2fl6czMrA01u1tsMDBE0mBgY2ApsC9wWR4/HZiUX0/Mw+TxEySpyfGZmVkTNC25RMQ/gG8BC0lJZSUwD3g8IlbnyRYDI/LrEcCiPO/qPP3WlfVKmipprqS5K1asaFb4ZmbWC83sFtuS1BrZAdgO2ATYv8qk0TlLnXFrCiLOi4hxETGuo6Opt4A2M7Meama32DuBv0fEioh4HvgF8C/A0NxNBjASWJJfLwZGAeTxWwCPNjE+MzNrkmYml4XAeEkb52MnE4C/ANcCB+dpJgMz8+tZeZg8/pqIWKflYmZm/V8zj7nMIR2YvwW4My/rPOB44DhJ80nHVKblWaYBW+fy44ATmhWbmZk1V1MvXBkRJwEnVRQ/AOxZZdpngUOaGY+ZmfUN/0PfzMxK5+RiZmalc3IxM7PSObmYmVnpnFzMzKx0Ti5mZlY6JxczMyudk4uZmZXOycXMzErn5GJmZqVzcjEzs9I5uZiZWemcXMzMrHROLmZmVjonFzMzK52Ti5mZla5pyUXSLpJuKzyekHSspK0kXS3pvvy8ZZ5eks6WNF/SHZJ2b1ZsZmbWXM28zfG9EbFbROwGvAl4GricdPvi2RExFpjNmtsZ7w+MzY+pwLnNis3MzJqrr7rFJgD3R8SDwERgei6fDkzKrycCF0ZyEzBU0vA+is/MzErUV8nlUODi/HrbiFgKkJ+3yeUjgEWFeRbnsrVImipprqS5K1asaGLIZmbWU01PLpI2AN4P/KyrSauUxToFEedFxLiIGNfR0VFGiGZmVrK+aLnsD9wSEcvy8LLO7q78vDyXLwZGFeYbCSzpg/jMzKxkfZFcDmNNlxjALGByfj0ZmFkoPzKfNTYeWNnZfWZmZu1lcDMrl7Qx8C7gE4Xi04FLJU0BFgKH5PIrgQOA+aQzy45qZmxmZtY8TU0uEfE0sHVF2SOks8cqpw3g6GbGY2ZmfcP/0Dczs9I5uZiZWemcXMzMrHROLmZmVjonFzMzK52Ti5mZlc7JxczMSufkYmZmpXNyMTOz0jm5mJlZ6ZxczMysdE4uZmZWOicXMzMrnZOLmZmVrlvJRdIWknZtVjBmZjYwdJlcJM2WtLmkLYE7gRmSvtn80MzMrF010nLZKiKeAA4CpkfEbsB7Gqlc0lBJl0n6q6R7JL1Z0laSrpZ0X37eMk8rSWdLmi/pDkm793y1zMyslRpJLoMldZBuR/yrbtb/HeC3EfEq4A3APcAJwOyIGAvMzsMA+wNj82MqcG43l2VmZv1EI8nl68D1wMKIuFnSjsDfu5pJ0ubA24BpABHxz4h4HJgITM+TTQcm5dcTgQsjuQkYKml4t9bGzMz6hcENTLMgIl46iB8RD0g6rYH5dgRWAP8n6Q3APOAYYNuIWJrrWippmzz9CGBRYf7FuWxpsVJJU0ktG0aPHt1AGGZm1tcaabmcU6Xsew3MNxjYHTg3It4IPMWaLrBqVKUs1imIOC8ixkXEuI6OjgbCMDOzvlaz5SJpT+DNQIekzxVGbQ6s30Ddi4HFETEnD19GSi7LJA3PrZbhwPLC9KMK848EljS2GmZm1p/Ua7lsAgwjJaCOwuOfpIP7dUXEQ8AiSbvkognAX4BZwORcNhmYmV/PAo7MZ42NB1Z2dp+ZmVl7qdlyiYhrgWsl/V8+zrJhRDzXzfo/C/xE0gbAA8BRpIR2qaQpwELWJKorgQOA+cDTeVozM2tDjRzQHyZpJrAFMDofnP9YRHy2qxkj4jZgXJVRE6pMG8DRDcRjZmb9XCMH9M8GDgQeAYiI24F3NDMoMzNrb40kl/Ui4sGKsheaEYyZmQ0MjXSLLcpnjoWkQaTjKH9rblhmZtbOGmm5fAo4DhgNLAPG5zIzM7Oqumy5RMRy4NA+iMXMzAaILpOLpDOrFK8E5kbEr8sPyczM2l0j3WKbAXuRrvu1CNgDeAXwaUlnNDE2MzNrU40c0N8J2CcingeQ9F3gt6R7utwOfKF54ZmZWTtqpOUyAhhSGB4CjIiI1UB3/7FvZmYvA420XM4EbpM0m3Tl4n2Ab0raBLiueaGZmVm7qptcJIl0Qclfk467CDglIjrvu3Jcc8MzM7N2VDe5RERIuiIi3kS6JL6ZmVmXGjnmcrOk3ZseiZmZDRiNHHN5C/BxSfeT7iYpUqPGCcfMzKpqJLlManoUZmY2oHTZLRYR90fE/cBjwDOFR5ckLZB0p6TbJM3NZVtJulrSffl5y1wuSWdLmi/pDnfFmZm1ry6Ti6T3Svob6YD+HNK/9K/pxjLeERG7RUTnTcNOAGZHxFhgdh4G2B8Ymx9TgXO7sQwzM+tHGukW+zqwN3BVRLxR0ruAD/RimRNJ/5UBmE76r8zxufzCfEfKmyQNlTQ8Ipb2YlkD1ow5C3s1/4f2Gl1SJGZm62rkbLHVEbECWE+SIuJqoNEuqwCukjRP0tRctm1nwsjP2+TyEaRWUafFuczMzNpMIy2Xlfnf+DcAF0paDrzYYP17R8QSSdsAV0v6a51pVaUs1pkoJampAKNH+9e3mVl/1EjLZRLwLHAsqQvrH8D7Gqk8Ipbk5+XA5cCewDJJwwHy8/I8+WJgVGH2kcCSKnWeFxHjImJcR0dHI2GYmVkfq5lcJF0FEBGrIuKFiHg+IqZFxJm5m6wuSZtI2qzzNfBu4C7S5WQm58kmAzPz61nAkfmssfHASh9vMTNrT/W6xXrbLNgWuDxdnozBwIyI+K2kPwOXSpoCLAQOydNfCRwAzAeeBo7q5fLNzKxF6iWXLSQdVGtkRPyiXsUR8QDwhirljwATqpQHcHS9Os3MrD3UTS7AgdQ+0F43uZiZ2ctXveTyYER8tM8iMTOzAaPe2WLVWixmZmZdqpdcjuizKMzMbECpmVwi4q6+DMTMzAaORv5EaWZm1i31/kQ5Oz9/o+/CMTOzgaDe2WLDJb0deL+kS6g4wB8RtzQ1MjMza1v1kstXSfdaGQmcWTEugH2bFZSZmbW3msklIi4DLpP0nxFxah/GZGZmba7LS+5HxKmS3g+8LRddFxFXNDcsMzNrZ43c5vg04BjgL/lxTC4zMzOrqpGbhb0X2C0iXgSQNB24FfhyMwMzM7P21ej/XIYWXm/RjEDMzGzgaKTlchpwq6RrSacjvw23WszMrI5GDuhfLOk6YA9Scjk+Ih5qdmBmZta+GuoWi4ilETErImZ2N7FIGiTpVklX5OEdJM2RdJ+kn0raIJdvmIfn5/FjursyZmbWP/TFtcWOAe4pDH8DOCsixgKPAVNy+RTgsYjYGTgrT2dmZm2oqclF0kjS2Wbn52GR/tl/WZ5kOjApv56Yh8njJ+TpzcyszdRNLpLWk9SbS+9/G/gS8GIe3hp4PCJW5+HFwIj8egSwCCCPX5mnr4xpqqS5kuauWLGiF6GZmVmz1E0u+b8tt0sa3d2KJR0ILI+IecXiaotpYFwxpvMiYlxEjOvo6OhuWGZm1gcaORV5OHC3pJuBpzoLI+L9Xcy3N+mKygcAGwGbk1oyQyUNzq2TkcCSPP1iYBSwWNJg0v9pHu3OypiZWf/QSHI5pScVR8SXyf+HkbQP8MWIOFzSz4CDgUuAycDMPMusPHxjHn9NRKzTcjEzs/6vkf+5XC9pe2BsRPxe0sbAoF4s83jgEklfI11GZlounwZcJGk+qcVyaC+WYWZmLdRlcpH0cWAqsBWwE+nA+/eBCY0uJCKuA67Lrx8A9qwyzbPAIY3WaWZm/VcjpyIfTTp+8gRARNwHbNPMoMzMrL01klyei4h/dg7kg+0+FmJmZjU1klyul/QVYIikdwE/A37V3LDMzKydNZJcTgBWAHcCnwCuBE5sZlBmZtbeGjlb7MV8g7A5pO6we32KsJmZ1dPI2WLvJZ0ddj/pX/Q7SPpERPym2cGZmVl7auRPlGcA74iI+QCSdgJ+DTi5mJlZVY0cc1nemViyB4DlTYrHzMwGgJotF0kH5Zd3S7oSuJR0zOUQ4M99EJuZmbWpet1i7yu8Xga8Pb9eAWzZtIjMzKzt1UwuEXFUXwZiZmYDRyNni+0AfBYYU5y+gUvum5nZy1QjZ4v9knTF4l+x5o6SZmZmNTWSXJ6NiLObHomZmQ0YjSSX70g6CbgKeK6zMCJuaVpUZmbW1hpJLq8DjgD2ZU23WORhMzOzdTSSXP4V2LF42f1GSNoI+AOwYV7OZRFxUj5B4BLSzcduAY6IiH9K2hC4EHgT8AjwwYhY0J1lWvuYMWdhr+b/0F6jS4rEzJqhkX/o3w4M7UHdzwH7RsQbgN2A/SSNB74BnBURY4HHgCl5+inAYxGxM3BWns7MzNpQI8llW+Cvkn4naVbno6uZInkyD66fH53daZfl8unApPx6Yh4mj58gSQ2uh5mZ9SONdIud1NPKJQ0C5gE7A98jXVn58YhYnSdZDIzIr0cAiwAiYrWklcDWwMMVdU4FpgKMHu2uETOz/qiR+7lc39PKI+IFYDdJQ4HLgVdXmyw/V2ulrHPfmIg4DzgPYNy4cb6vjJlZP9Rlt5ikVZKeyI9nJb0g6YnuLCQiHgeuA8YDQyV1JrWRwJL8ejEwKi9zMLAF8Gh3lmNmZv1Dl8klIjaLiM3zYyPgA8B3u5pPUkdusSBpCPBO4B7gWuDgPNlkYGZ+PSsPk8df4ztempm1p0aOuawlIn4p6YQGJh0OTM/HXdYDLo2IKyT9BbhE0teAW0mXliE/XyRpPqnFcmh3YzMzs/6hkQtXHlQYXA8YR5VjIZUi4g7gjVXKHwD2rFL+LOleMWZm1uYaabkU7+uyGlhAOm3YzMysqkbOFvN9XczMrFvq3eb4q3Xmi4g4tQnxmJnZAFCv5fJUlbJNSJdp2RpwcjEzs6rq3eb4jM7XkjYDjgGOIl108oxa85mZmdU95iJpK+A44HDSdb92j4jH+iIwMzNrX/WOuXwTOIh0qZXXFS5CaWZmVle9f+h/AdgOOBFYUrgEzKruXv7FzMxeXuodc2nkcvxmZmbrcAIxM7PSObmYmVnpnFzMzKx0Ti5mZlY6JxczMyudk4uZmZXOycXMzErXtOQiaZSkayXdI+luScfk8q0kXS3pvvy8ZS6XpLMlzZd0h6TdmxWbmZk1VzNbLquBL0TEq4HxwNGSdgVOAGZHxFhgdh4G2B8Ymx9TgXObGJuZmTVR05JLRCyNiFvy61XAPcAI0l0sp+fJpgOT8uuJwIWR3AQMlTS8WfGZmVnz9MkxF0ljgDcCc4BtI2IppAQEbJMnGwEsKsy2OJdV1jVV0lxJc1esWNHMsM3MrIeanlwkbQr8HDg2Iupd8FJVymKdgojzImJcRIzr6OgoK0wzMytRU5OLpPVJieUnEfGLXLyss7srPy/P5YuBUYXZRwJLmhmfmZk1RzPPFhMwDbgnIs4sjJoFTM6vJwMzC+VH5rPGxgMrO7vPzMysvdS9E2Uv7Q0cAdwp6bZc9hXgdOBSSVOAhcAhedyVwAHAfOBp0i2VzcysDTUtuUTEDVQ/jgIwocr0ARzdrHjMzKzv+B/6ZmZWumZ2i5kNWDPmLOzV/B/aa3RJkZj1T265mJlZ6ZxczMysdE4uZmZWOicXMzMrnZOLmZmVzsnFzMxK5+RiZmalc3IxM7PSObmYmVnpnFzMzKx0Ti5mZlY6JxczMyudk4uZmZXOycXMzErXzNsc/0jSckl3Fcq2knS1pPvy85a5XJLOljRf0h2Sdm9WXGZm1nzNbLlcAOxXUXYCMDsixgKz8zDA/sDY/JgKnNvEuMzMrMmallwi4g/AoxXFE4Hp+fV0YFKh/MJIbgKGShrerNjMzKy5+vqYy7YRsRQgP2+Ty0cAiwrTLc5l65A0VdJcSXNXrFjR1GDNzKxn+ssBfVUpi2oTRsR5ETEuIsZ1dHQ0OSwzM+uJvk4uyzq7u/Lz8ly+GBhVmG4ksKSPYzMzs5L0dXKZBUzOrycDMwvlR+azxsYDKzu7z8zMrP0MblbFki4G9gGGSVoMnAScDlwqaQqwEDgkT34lcAAwH3gaOKpZcZmZWfM1LblExGE1Rk2oMm0ARzcrFjMz61v95YC+mZkNIE1ruZhZ/zZjzsJezf+hvUaXFIkNRG65mJlZ6ZxczMysdE4uZmZWOicXMzMrnZOLmZmVzsnFzMxK5+RiZmalc3IxM7PSObmYmVnpnFzMzKx0Ti5mZlY6X1vMzNqSr43Wv7nlYmZmpXNyMTOz0vWr5CJpP0n3Spov6YRWx2NmZj3Tb5KLpEHA94D9gV2BwyTt2tqozMysJ/rTAf09gfkR8QCApEuAicBfWhqVmVkTDPQTEpRuX996kg4G9ouIj+XhI4C9IuIzFdNNBabmwV2Ae3u4yGHAwz2ct7/xuvQ/A2U9wOvSX/VmXbaPiI4yg6nUn1ouqlK2TuaLiPOA83q9MGluRIzrbT39gdel/xko6wFel/6qv69LvznmAiwGRhWGRwJLWhSLmZn1Qn9KLn8GxkraQdIGwLVZzJcAAAitSURBVKHArBbHZGZmPdBvusUiYrWkzwC/AwYBP4qIu5u4yF53rfUjXpf+Z6CsB3hd+qt+vS795oC+mZkNHP2pW8zMzAYIJxczMyvdyy65SPqRpOWS7mp1LL0laZSkayXdI+luSce0OqaekLSRpJsl3Z7X45RWx9RbkgZJulXSFa2OpTckLZB0p6TbJM1tdTw9JWmopMsk/TV/X97c6ph6QtIu+b3ofDwh6dhWx1XNy+6Yi6S3AU8CF0bEa1sdT29IGg4Mj4hbJG0GzAMmRURbXdVAkoBNIuJJSesDNwDHRMRNLQ6txyQdB4wDNo+IA1sdT09JWgCMi4i2/uOhpOnAHyPi/Hw26sYR8Xir4+qNfMmsf5D+bP5gq+Op9LJruUTEH4BHWx1HGSJiaUTckl+vAu4BRrQ2qu6L5Mk8uH5+tO2vHkkjgfcC57c6FgNJmwNvA6YBRMQ/2z2xZBOA+/tjYoGXYXIZqCSNAd4IzGltJD2Tu5FuA5YDV0dEW65H9m3gS8CLrQ6kBAFcJWlevvRSO9oRWAH8X+6qPF/SJq0OqgSHAhe3OohanFwGAEmbAj8Hjo2IJ1odT09ExAsRsRvpygx7SmrLLktJBwLLI2Jeq2Mpyd4RsTvpauVH527ldjMY2B04NyLeCDwFtPUtPXLX3vuBn7U6llqcXNpcPkbxc+AnEfGLVsfTW7m74jpgvxaH0lN7A+/PxyouAfaV9OPWhtRzEbEkPy8HLiddvbzdLAYWF1rDl5GSTTvbH7glIpa1OpBanFzaWD4QPg24JyLObHU8PSWpQ9LQ/HoI8E7gr62Nqmci4ssRMTIixpC6La6JiA+3OKwekbRJPlGE3I30bqDtzrKMiIeARZJ2yUUTaP9beRxGP+4Sg350+Ze+IuliYB9gmKTFwEkRMa21UfXY3sARwJ35eAXAVyLiyhbG1BPDgen57Jf1gEsjoq1P4R0gtgUuT79hGAzMiIjftjakHvss8JPcnfQAcFSL4+kxSRsD7wI+0epY6nnZnYpsZmbN524xMzMrnZOLmZmVzsnFzMxK5+RiZmalc3IxM7PSOblYr0gKSWcUhr8o6eSS6r5A0sFl1NXFcg7JV8q9thvzXCfp3nwl5z9L2q0wboGknxeGD5Z0QcX8MyXdWMoKNBbvsfkU1mYvZ4ykDzV7Odb/OblYbz0HHCRpWKsDKcr/mWnUFODTEfGObi7m8Ih4A3AO8M2KceMkvaZGbENJ/xAfKmmHbi6zS0oqv9vHAt1KLt3chp3GAE4u5uRivbaadC/vz1eOqGx5SHoyP+8j6XpJl0r6m6TTJR2e7+lyp6SdCtW8U9If83QH5vkHSfpmbjHcIekThXqvlTQDuLNKPIfl+u+S9I1c9lXgLcD3JX2zYvr1JJ2jdI+ZKyRdWaMldSPrXo36W8BXamyzDwC/Il0e5tBqE0g6WdJFkq6RdJ+kjxfG/Xth3U/JZWNy6+sc4BZgVGH6zwHbAdd2ts4knStprirun5NbXV+VdANwiKQ98nJuzNv8rjxd1fcAOB14q9K9Rj4v6TX5fb0tTze2xjaxgSYi/PCjxw/SvXE2BxYAWwBfBE7O4y4ADi5Om5/3AR4n/TN/Q9I9KU7J444Bvl2Y/7ekH0FjSdeI2giYCpyYp9kQmAvskOt9CtihSpzbAQuBDtK/za8h3fsG0rXMxlWZ52Dgyrz8VwCPda5PcR5Sq+C/C/MtIP27/R5g51zPBYXxvwfeCrwSuKPGdj0ZuB0YAgwDFuV1eDcpmSvHdQXpcvJjSFdhHl+jvgXAsMLwVvl5UF6X1xem+1JhuruAf8mvTwfuyq/rvQdXFOb/X1ILD2ADYEirP7N+9M3jZXf5FytfRDwh6ULgc8AzDc7254hYCiDpfuCqXH4nUOyeujQiXgTuk/QA8CrSDvb1hVbEFqTk80/g5oj4e5Xl7QFcFxEr8jJ/Qtop/7JOjG8BfpaX/1CVYzI/ydfcGsS6F0J8gdRV9mXgN52FkrYlJZwbIiIkrZb02oiods2umRHxDPBMXvaeOaZ3A7fmaTbN674QeDAav8HavyldQn8wKcnvCtyRx/00xzoU2Cwi/pTLZwCdNz6r9x4U3Qj8h9I9bn4REfc1GJ+1OXeLWVm+TTp2UbxPxmryZ0zpAlUbFMY9V3j9YmH4Rda+5l3l9YmC9Kv9sxGxW37sEBGdyempGvGp0RXpxjyHk36tzwC+V2X8RaQENrpQ9kFgS+DvSldOHkONrjFqr/tphXXfOdZcG6/Wuq8lH+f5IjAhIl4P/JrUIuzUWU+99a/3HqwJOGIG6dLwzwC/k7RvIzFa+3NysVJExKPApaQE02kB8Kb8eiLpDpPddUg+9rET6aZP9wK/Az6ldLsBJL1SXd/8aQ7wdknD8oHqw4Dru5jnBuADefnbkrp81hIRzwMnAuMlvbrKuLNI3WadDgP2i4gxka6c/CZqJ5eJkjaStHVe9p9J6/5RpXv4IGmEpG26WA+AVcBm+fXmpASyMq/X/tVmiIjHgFWSxueiYpy13oPicpC0I/BARJwNzAJe30CsNgC4W8zKdAbwmcLwD4GZkm4GZtPgL+sK95KSwLbAJyPiWUnnk37x35JbRCuASfUqiYilkr4MXEv61X1lRMzsYtk/J12e/S7gb6QEtbJK3c8onY79RdZOrpBuiXAivHS30NHATYV5/y7pCUl7xbp337yZ1KoYDZwa6d4qS3ISuzGtOk8CHyZ1w9VzHvAbSUsj4h2SbgXuJl0h+P/VmW8K8ENJT5GOzXSuf6334A5gtaTbScfMNgI+LOl54CHgv7qI0wYIXxXZrA5Jm0bEk7n1cDPpzowP9cFyTyadAPGtZi+rizg2jYjOs/xOAIZHxDGtjMnag1suZvVdkQ9sb0BqPTQ9sfQz780tvsHAg8BHWhuOtQu3XMzMrHQ+oG9mZqVzcjEzs9I5uZiZWemcXMzMrHROLmZmVrr/Dz/4Fx0VFwbYAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "ax = sns.distplot([len(exp_tid_gids_map[tid]) for tid in exp_tid_gids_map], kde=False)\n",
    "ax.set_title('Distribution of number of gRNAs that target a splice acceptor')\n",
    "ax.set_ylabel('Number of Targets')\n",
    "ax.set_xlabel('Number of gRNA per targets')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Barcode - Sequence Extraction\n",
    "We have 3 categories of sequence reads (with replicates): GDNA, 'PreCas-Unspliced' or 'PostCas-Unspliced', and 'PreCas-Spliced' or 'PostCas-Spliced'.\n",
    "\n",
    "## GDNA\n",
    "Captures the barcode sequence and gRNA sequence from the genomic DNA.\n",
    "\n",
    "## Pre/PostCas-Unspliced\n",
    "Captures the Exon B portion of the target sequence and the barcode sequence from the RNA primary transcript.\n",
    "\n",
    "## Pre/PostCas-Spliced\n",
    "Captures the sequence between the end of Exon A and the barcode sequence (we can check whether Exon B is present or not in this range).\n",
    "\n",
    "## What the following code does\n",
    "The sequence (gRNA sequence in GDNA, the target sequence in Unspliced, the Exon B portion of the target sequence, or lack thereof, in Spliced), the associated barcode sequences, and the sequenc are extracted from the sequencing reads using regular expression pattern matching with at most 5 substitutions permitted (see code comments). The results are written to disk for subsequent processing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_pairs_iterator(suffix):\n",
    "    read1_fn = os.path.join(READS_DIR, READ_FN_PREFIX + suffix + '_1_sequence.fastq')\n",
    "    read2_fn = os.path.join(READS_DIR, READ_FN_PREFIX + suffix + '_2_sequence.fastq')\n",
    "    with open(read1_fn, 'r') as read1_handle:\n",
    "        with open(read2_fn, 'r') as read2_handle:\n",
    "            read1_records = SeqIO.parse(read1_handle, \"fastq-sanger\")\n",
    "            read2_records = SeqIO.parse(read2_handle, \"fastq-sanger\")\n",
    "            while 1:\n",
    "                try:\n",
    "                    read1 = next(read1_records)\n",
    "                    read2 = next(read2_records)\n",
    "                    yield (read1, read2)\n",
    "                except StopIteration as e:\n",
    "                        break\n",
    "\n",
    "                        \n",
    "def get_line_count(fn):\n",
    "        with open(fn) as f:\n",
    "            return sum(1 for line in f)\n",
    "        \n",
    "\n",
    "def get_total_read_pairs(suffix):\n",
    "    read1_fn = os.path.join(READS_DIR, READ_FN_PREFIX + suffix + '_1_sequence.fastq')\n",
    "    read2_fn = os.path.join(READS_DIR, READ_FN_PREFIX + suffix + '_2_sequence.fastq')\n",
    "    return int(min(get_line_count(read1_fn), get_line_count(read2_fn)) / 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Barcode - gRNA Seq Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_bc_grna(gdna_suffix, output_fn):\n",
    "    # If already cached, do nothing\n",
    "    if os.path.exists(output_fn):\n",
    "        return\n",
    "    \n",
    "    total_pairs = get_total_read_pairs(gdna_suffix)\n",
    "    \n",
    "    # Read 1 pattern match (gRNA) [fuzzy match allowing MAX_GDNA_R1_SUBSTITUTIONS substitutions]:\n",
    "    # GACGAAACACCG(gRNA seq: N*20 or N*19)GTTTAAGAGCTA\n",
    "    r1_pattern_prefix = 'GACGAAACACCG'\n",
    "    r1_pattern_suffix = 'GTTTAAGAGCTA'\n",
    "    r1_regex = r1_pattern_prefix + '.{19,20}' + r1_pattern_suffix\n",
    "    r1_fuzzy = '{s<=' + str(MAX_GDNA_R1_SUBSTITUTIONS) + '}'\n",
    "    r1_prog = regex.compile('('+r1_regex+')' + r1_fuzzy)\n",
    "    \n",
    "    # Read 2 pattern match (BC) [fuzzy match allowing MAX_GDNA_R2_SUBSTITUTIONS substitutions]:\n",
    "    # GTGCAGCTCGCC(rev comp BC seq: N*15)GACCACTACCAG\n",
    "    r2_pattern_prefix = 'GTGCAGCTCGCC'\n",
    "    r2_pattern_suffix = 'GACCACTACCAG'\n",
    "    r2_regex = r2_pattern_prefix + '.{15}' + r2_pattern_suffix\n",
    "    r2_fuzzy = '{s<=' + str(MAX_GDNA_R2_SUBSTITUTIONS) + '}'\n",
    "    r2_prog = regex.compile('('+r2_regex+')' + r2_fuzzy)\n",
    "    \n",
    "    total_accepted = 0\n",
    "    fuzzy_matches = 0\n",
    "    with open(output_fn, 'w', newline='') as csvfile:\n",
    "        csvwriter = csv.writer(csvfile)\n",
    "        for r1, r2 in tqdm(read_pairs_iterator(gdna_suffix), total=total_pairs):\n",
    "            r1_seq = str(r1.seq).upper()\n",
    "            r2_seq = str(r2.seq).upper()\n",
    "\n",
    "            # Pattern match\n",
    "            m1 = r1_prog.search(r1_seq)\n",
    "            m2 = r2_prog.search(r2_seq)\n",
    "            if m1 is None or m2 is None:\n",
    "                continue\n",
    "            if m1.fuzzy_counts[0] > 0 or m2.fuzzy_counts[0] > 0:\n",
    "                fuzzy_matches += 1\n",
    "\n",
    "            g_start, g_end = m1.start() + len(r1_pattern_prefix), m1.end() - len(r1_pattern_suffix)\n",
    "            bc_start, bc_end = m2.start() + len(r2_pattern_prefix), m2.end() - len(r2_pattern_suffix)\n",
    "\n",
    "            g_quality = quality_scores_to_string(r1.letter_annotations[\"phred_quality\"][g_start:g_end])\n",
    "            bc_quality = quality_scores_to_string(r2.letter_annotations[\"phred_quality\"][bc_start:bc_end])\n",
    "                \n",
    "            g_seq = r1_seq[g_start:g_end]\n",
    "            bc_seq = reverse_complement(r2_seq[bc_start:bc_end])\n",
    "            csvwriter.writerow([bc_seq, bc_quality, g_seq, g_quality])\n",
    "            total_accepted += 1\n",
    "\n",
    "    stats = {\n",
    "        'exact_accepted': total_accepted - fuzzy_matches,\n",
    "        'fuzzy_accepted': fuzzy_matches,\n",
    "        'total_accepted': total_accepted,\n",
    "        'total': total_pairs\n",
    "    }\n",
    "\n",
    "    print(\"Exact Matches:\", stats['exact_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['exact_accepted'] * 100 / stats['total']))\n",
    "    print(\"Fuzzy Matches:\", stats['fuzzy_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['fuzzy_accepted'] * 100 / stats['total']))\n",
    "    print(\"Total Accepted:\", stats['total_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['total_accepted'] * 100 / stats['total']))\n",
    "    \n",
    "    return stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "identify_bc_grna(PRECAS_GDNA_SUFFIX, PRECAS_GDNA_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 4804805 out of 6469306 (74.27%)\n",
    "# Fuzzy Matches: 908110 out of 6469306 (14.04%)\n",
    "# Total Accepted: 5712915 out of 6469306 (88.31%)\n",
    "\n",
    "identify_bc_grna(POSTCAS_GDNA1_SUFFIX, POSTCAS_GDNA1_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 3846121 out of 5432654 (70.80%)\n",
    "# Fuzzy Matches: 1029915 out of 5432654 (18.96%)\n",
    "# Total Accepted: 4876036 out of 5432654 (89.75%)\n",
    "\n",
    "identify_bc_grna(POSTCAS_GDNA2_SUFFIX, POSTCAS_GDNA2_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 2560503 out of 3332188 (76.84%)\n",
    "# Fuzzy Matches: 472406 out of 3332188 (14.18%)\n",
    "# Total Accepted: 3032909 out of 3332188 (91.02%)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Barcode - PreCas Unspliced Target Seq Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def identify_bc_unspliced_target(unspliced_suffix, output_fn, len_limit=None):\n",
    "    # If already cached, do nothing\n",
    "    if os.path.exists(output_fn):\n",
    "        return\n",
    "    \n",
    "    total_pairs = get_total_read_pairs(unspliced_suffix)\n",
    "    \n",
    "    # Read 1 pattern match (target) [fuzzy match allowing MAX_UNSPLICED_R1_SUBSTITUTIONS substitutions]:\n",
    "    # GCCCTCGTCGAC(target seq: N*61)\n",
    "    # Most reads are too short to have anything beyond the 61bp target seq\n",
    "    r1_pattern_prefix = 'GCCCTCGTCGAC'\n",
    "    r1_regex = r1_pattern_prefix\n",
    "    r1_fuzzy = '{s<=' + str(MAX_UNSPLICED_R1_SUBSTITUTIONS) + '}'\n",
    "    r1_prog = regex.compile('('+r1_regex+')' + r1_fuzzy)\n",
    "    \n",
    "    # Read 2 pattern match (BC) [fuzzy match allowing MAX_UNSPLICED_R2_SUBSTITUTIONS substitutions]:\n",
    "    # CTGGTAGTGGTC(BC seq: N*15)GGCGAGCTGCAC\n",
    "    r2_pattern_prefix = 'CTGGTAGTGGTC'\n",
    "    r2_pattern_suffix = 'GGCGAGCTGCAC'\n",
    "    r2_regex = r2_pattern_prefix + '.{15}' + r2_pattern_suffix\n",
    "    r2_fuzzy = '{s<=' + str(MAX_UNSPLICED_R2_SUBSTITUTIONS) + '}'\n",
    "    r2_prog = regex.compile('('+r2_regex+')' + r2_fuzzy)\n",
    "    \n",
    "    len_limit = len_limit if len_limit is not None else 1000\n",
    "    \n",
    "    total_accepted = 0\n",
    "    fuzzy_matches = 0\n",
    "    with open(output_fn, 'w', newline='') as csvfile:\n",
    "        csvwriter = csv.writer(csvfile)\n",
    "        for r1, r2 in tqdm(read_pairs_iterator(unspliced_suffix), total=total_pairs):\n",
    "            r1_seq = str(r1.seq).upper()\n",
    "            r2_seq = str(r2.seq).upper()\n",
    "            \n",
    "            # Pattern match\n",
    "            m1 = r1_prog.search(r1_seq)\n",
    "            m2 = r2_prog.search(r2_seq)\n",
    "            if m1 is None or m2 is None:\n",
    "                continue\n",
    "            if m1.fuzzy_counts[0] > 0 or m2.fuzzy_counts[0] > 0:\n",
    "                fuzzy_matches += 1\n",
    "                \n",
    "            t_start = m1.start() + len(r1_pattern_prefix)\n",
    "            t_end = t_start + len_limit\n",
    "            bc_start, bc_end = m2.start() + len(r2_pattern_prefix), m2.end() - len(r2_pattern_suffix)\n",
    "\n",
    "            t_quality = quality_scores_to_string(r1.letter_annotations[\"phred_quality\"][t_start:t_end])\n",
    "            bc_quality = quality_scores_to_string(r2.letter_annotations[\"phred_quality\"][bc_start:bc_end])\n",
    "            \n",
    "            t_seq = r1_seq[t_start:t_end]\n",
    "            bc_seq = r2_seq[bc_start:bc_end]\n",
    "            csvwriter.writerow([bc_seq, bc_quality, t_seq, t_quality])\n",
    "            total_accepted += 1\n",
    "            \n",
    "    stats = {\n",
    "        'exact_accepted': total_accepted - fuzzy_matches,\n",
    "        'fuzzy_accepted': fuzzy_matches,\n",
    "        'total_accepted': total_accepted,\n",
    "        'total': total_pairs\n",
    "    }\n",
    "\n",
    "    print(\"Exact Matches:\", stats['exact_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['exact_accepted'] * 100 / stats['total']))\n",
    "    print(\"Fuzzy Matches:\", stats['fuzzy_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['fuzzy_accepted'] * 100 / stats['total']))\n",
    "    print(\"Total Accepted:\", stats['total_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['total_accepted'] * 100 / stats['total']))\n",
    "    \n",
    "    return stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "precas_unspliced_stats = identify_bc_unspliced_target(PRECAS_UNSPLICED_SUFFIX, PRECAS_UNSPLICED_BC_SEQ, 61)\n",
    "# Stats:\n",
    "# Exact Matches: 13994894 out of 17793229 (78.65%)\n",
    "# Fuzzy Matches: 1007851 out of 17793229 (5.66%)\n",
    "# Total Accepted: 15002745 out of 17793229 (84.32%)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Barcode - PostCas Unspliced Target Seq Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "identify_bc_unspliced_target(POSTCAS_UNSPLICED1_SUFFIX, POSTCAS_UNSPLICED1_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 19867537 out of 24052905 (82.60%)\n",
    "# Fuzzy Matches: 1424870 out of 24052905 (5.92%)\n",
    "# Total Accepted: 21292407 out of 24052905 (88.52%)\n",
    "\n",
    "identify_bc_unspliced_target(POSTCAS_UNSPLICED2_SUFFIX, POSTCAS_UNSPLICED2_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 21579179 out of 25490883 (84.65%)\n",
    "# Fuzzy Matches: 1566151 out of 25490883 (6.14%)\n",
    "# Total Accepted: 23145330 out of 25490883 (90.80%)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Barcode - PreCas Spliced Target Seq Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_spliced_BC(r2m, r1m):\n",
    "    r2, r2_seq, m2, r2_pattern_prefix, r2_pattern_suffix = r2m\n",
    "    r1, r1_seq, m1, r1_pattern_prefix, r1_pattern_suffix = r1m\n",
    "    \n",
    "    if m1:\n",
    "        bc1_start, bc1_end = m1.start() + len(r1_pattern_prefix), m1.end() - len(r1_pattern_suffix)\n",
    "        bc1_quality = r1.letter_annotations[\"phred_quality\"][bc1_start:bc1_end][::-1] # reverse!\n",
    "        bc1_seq = reverse_complement(r1_seq[bc1_start:bc1_end])\n",
    "    if m2:\n",
    "        bc2_start, bc2_end = m2.start() + len(r2_pattern_prefix), m2.end() - len(r2_pattern_suffix)\n",
    "        bc2_quality = r2.letter_annotations[\"phred_quality\"][bc2_start:bc2_end]\n",
    "        bc2_seq = r2_seq[bc2_start:bc2_end]\n",
    "        \n",
    "    if m1 and not m2:\n",
    "        return bc1_seq, bc1_quality\n",
    "    if m2 and not m1:\n",
    "        return bc2_seq, bc2_quality\n",
    "    if m1 and m2:\n",
    "        if bc1_seq == bc2_seq:\n",
    "            return bc1_seq, bc1_quality if sum(bc1_quality) > sum(bc2_quality) else bc2_quality\n",
    "\n",
    "        \n",
    "        # Only consider pairs with at most 3 mismatches\n",
    "        if hamming_distance(bc1_seq, bc2_seq) <= 3:\n",
    "            merged_bc_seq = []\n",
    "            merged_bc_quality = []\n",
    "            for i in range(len(bc1_seq)):\n",
    "                if bc1_seq[i] == bc2_seq[i]:\n",
    "                    merged_bc_seq.append(bc1_seq[i])\n",
    "                    merged_bc_quality.append(max(bc1_quality[i], bc2_quality[i]))\n",
    "                else:\n",
    "                    if bc1_quality[i] > bc2_quality[i]:\n",
    "                        merged_bc_seq.append(bc1_seq[i])\n",
    "                        merged_bc_quality.append(bc1_quality[i])\n",
    "                    elif bc2_quality[i] > bc1_quality[i]:\n",
    "                        merged_bc_seq.append(bc2_seq[i])\n",
    "                        merged_bc_quality.append(bc2_quality[i])\n",
    "                    else:\n",
    "                        # tie break by seeing which bc sequence has overall higher quality\n",
    "                        overall_bc1_quality, overall_bc2_quality = sum(bc1_quality), sum(bc2_quality)\n",
    "                        if overall_bc1_quality > overall_bc2_quality:\n",
    "                            merged_bc_seq.append(bc1_seq[i])\n",
    "                            merged_bc_quality.append(bc1_quality[i])\n",
    "                        elif overall_bc1_quality > overall_bc2_quality:\n",
    "                            merged_bc_seq.append(bc2_seq[i])\n",
    "                            merged_bc_quality.append(bc2_quality[i])\n",
    "                        else:\n",
    "                            # give up\n",
    "                            return None, None\n",
    "\n",
    "            assert(len(merged_bc_seq) == 15)\n",
    "            bc_seq = ''.join(merged_bc_seq)\n",
    "            return bc_seq, merged_bc_quality\n",
    "    return None, None\n",
    "    \n",
    "\n",
    "def identify_bc_spliced(spliced_suffix, output_fn):\n",
    "    # If already cached, do nothing\n",
    "    if os.path.exists(output_fn):\n",
    "        return\n",
    "    \n",
    "    total_pairs = get_total_read_pairs(spliced_suffix)    \n",
    "    \n",
    "    # Read 1 pattern match\n",
    "    # If exon B NOT skipped, match only target seq post AG, due to splicing:\n",
    "    # CACAACATCGAG(target seq post acceptor splice site (which should be AG))TGATTACACATA\n",
    "    #\n",
    "    # If exon B skipped, match exon C seq:\n",
    "    # CACAACATCGAG|GACGGCAGCGTGCAG\n",
    "    # As well as get the BC, and make sure it's the same as the BC in read2 by matching\n",
    "    # GTGCAGCTCGCC(rev comp BC seq: N*15)GACCACTACCAG\n",
    "    r1_pattern_prefix = 'CACAACATCGAG'\n",
    "    r1_pattern_suffix = 'TGATTACACATA'\n",
    "    r1_regex = r1_pattern_prefix + '.+' + r1_pattern_suffix\n",
    "    r1_fuzzy = '{s<=' + str(MAX_SPLICED_R1_SUBSTITUTIONS) + '}'\n",
    "    r1_prog = regex.compile('('+r1_regex+')' + r1_fuzzy)\n",
    "    \n",
    "    r1_SKIPPED_regex = 'CACAACATCGAGGACGGCAGCGTGCAG'\n",
    "    r1_SKIPPED_fuzzy = '{s<=' + str(MAX_SPLICED_SKP_R1_SUBSTITUTIONS) + '}'\n",
    "    r1_SKIPPED_prog = regex.compile('('+r1_SKIPPED_regex+')' + r1_SKIPPED_fuzzy)\n",
    "    \n",
    "    r1_SKIPPED_BC_pattern_prefix = 'GTGCAGCTCGCC'\n",
    "    r1_SKIPPED_BC_pattern_suffix = 'GACCACTACCAG'\n",
    "    r1_SKIPPED_BC_regex = r1_SKIPPED_BC_pattern_prefix + '.{15}' + r1_SKIPPED_BC_pattern_suffix\n",
    "    r1_SKIPPED_BC_fuzzy = '{s<=' + str(MAX_SPLICED_BC_R1_SUBSTITUTIONS) + '}'\n",
    "    r1_SKIPPED_BC_prog = regex.compile('('+r1_SKIPPED_BC_regex+')' + r1_SKIPPED_BC_fuzzy)\n",
    "    \n",
    "    # Read 2 pattern match (BC):\n",
    "    # CTGGTAGTGGTC(BC seq: N*15)GGCGAGCTGCAC\n",
    "    r2_pattern_prefix = 'CTGGTAGTGGTC'\n",
    "    r2_pattern_suffix = 'GGCGAGCTGCAC'\n",
    "    r2_regex = r2_pattern_prefix + '.{15}' + r2_pattern_suffix\n",
    "    r2_fuzzy = '{s<=' + str(MAX_SPLICED_R2_SUBSTITUTIONS) + '}'\n",
    "    r2_prog = regex.compile('('+r2_regex+')' + r2_fuzzy)\n",
    "    \n",
    "    total_accepted = 0\n",
    "    fuzzy_matches = 0\n",
    "    with open(output_fn, 'w', newline='') as csvfile:\n",
    "        csvwriter = csv.writer(csvfile)\n",
    "        for r1, r2 in tqdm(read_pairs_iterator(spliced_suffix), total=total_pairs):\n",
    "            r1_seq = str(r1.seq).upper()\n",
    "            r2_seq = str(r2.seq).upper()\n",
    "\n",
    "            m1 = r1_prog.search(r1_seq)\n",
    "            m1_skipped = r1_SKIPPED_prog.search(r1_seq)\n",
    "            m1_skipped_rc_BC = r1_SKIPPED_BC_prog.search(r1_seq)\n",
    "\n",
    "            m2 = r2_prog.search(r2_seq)\n",
    "\n",
    "            is_not_skipped = m1 and m1_skipped is None and m1_skipped_rc_BC is None\n",
    "            is_skipped = m1 is None and m1_skipped and m1_skipped_rc_BC\n",
    "\n",
    "            # Can't do much with these, probably sequencing errors preventing exact prefix/suffix matches\n",
    "            # among other reasons. Ignore these for now.\n",
    "            if not (is_not_skipped or is_skipped):\n",
    "                continue\n",
    "            # Can't be both at the same time...\n",
    "            if (is_not_skipped and is_skipped):\n",
    "                continue\n",
    "            \n",
    "            bc_seq, bc_quality = determine_spliced_BC((r2, r2_seq, m2, r2_pattern_prefix, r2_pattern_suffix),\n",
    "                                          (r1, r1_seq, m1_skipped_rc_BC, \n",
    "                                           r1_SKIPPED_BC_pattern_prefix, r1_SKIPPED_BC_pattern_suffix))\n",
    "            # BC must be decipherable\n",
    "            if not bc_seq:\n",
    "                continue\n",
    "                \n",
    "            if ((is_not_skipped and m1.fuzzy_counts[0] > 0) or \n",
    "                (is_skipped and (m1_skipped.fuzzy_counts[0] > 0 or m1_skipped_rc_BC.fuzzy_counts[0] > 0)) or \n",
    "                (m2 and m2.fuzzy_counts[0] > 0)):\n",
    "                fuzzy_matches += 1\n",
    "\n",
    "            if is_not_skipped:\n",
    "                t_start, t_end = m1.start() + len(r1_pattern_prefix), m1.end() - len(r1_pattern_suffix)\n",
    "                t_quality = quality_scores_to_string(r1.letter_annotations[\"phred_quality\"][t_start:t_end])\n",
    "    \n",
    "                t_seq = r1_seq[t_start:t_end]\n",
    "                total_accepted += 1\n",
    "                csvwriter.writerow([bc_seq, quality_scores_to_string(bc_quality), t_seq, t_quality])\n",
    "            elif is_skipped:\n",
    "                total_accepted += 1\n",
    "                csvwriter.writerow([bc_seq, quality_scores_to_string(bc_quality), 'SKIPPED'])\n",
    "\n",
    "    stats = {\n",
    "        'exact_accepted': total_accepted - fuzzy_matches,\n",
    "        'fuzzy_accepted': fuzzy_matches,\n",
    "        'total_accepted': total_accepted,\n",
    "        'total': total_pairs\n",
    "    }\n",
    "\n",
    "    print(\"Exact Matches:\", stats['exact_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['exact_accepted'] * 100 / stats['total']))\n",
    "    print(\"Fuzzy Matches:\", stats['fuzzy_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['fuzzy_accepted'] * 100 / stats['total']))\n",
    "    print(\"Total Accepted:\", stats['total_accepted'], \n",
    "          \"out of\", stats['total'], \n",
    "          '({:0.2f}%)'.format(stats['total_accepted'] * 100 / stats['total']))\n",
    "    \n",
    "    return stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "identify_bc_spliced(PRECAS_SPLICED_SUFFIX, PRECAS_SPLICED_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 3337696 out of 5902730 (56.54%)\n",
    "# Fuzzy Matches: 881806 out of 5902730 (14.94%)\n",
    "# Total Accepted: 4219502 out of 5902730 (71.48%)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Barcode - PostCas Spliced Target Seq Extraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "identify_bc_spliced(POSTCAS_SPLICED1_SUFFIX, POSTCAS_SPLICED1_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 4367086 out of 8380139 (52.11%)\n",
    "# Fuzzy Matches: 1108841 out of 8380139 (13.23%)\n",
    "# Total Accepted: 5475927 out of 8380139 (65.34%)\n",
    "\n",
    "identify_bc_spliced(POSTCAS_SPLICED2_SUFFIX, POSTCAS_SPLICED2_BC_SEQ)\n",
    "# Stats:\n",
    "# Exact Matches: 6370370 out of 9572842 (66.55%)\n",
    "# Fuzzy Matches: 1665108 out of 9572842 (17.39%)\n",
    "# Total Accepted: 8035478 out of 9572842 (83.94%)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:skipguide_data_processing] *",
   "language": "python",
   "name": "conda-env-skipguide_data_processing-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
